ext/OptimisersAdaptExt.jl

```diff
@@ -0,0 +1,18 @@
+module OptimisersAdaptExt
+
+import Adapt
+import Optimisers: Leaf
+
+function Adapt.adapt_structure(to, leaf::Leaf)
+  @warn """`Optimisers.Leaf` object does not support device transfer via
+  `Adapt.jl`. Avoid this by calling `Flux.gpu/cpu` or
+  `MLDataDevices.cpu_device()/gpu_device()` on the optimiser state object.
+  """ maxlog=1
```

Comment:
Do we want this only once? It's a potential correctness bug, not a performance issue. 
```suggestion
  """
```

Reply:
Is your worry that the message will get lost in the shuffle somehow? My thought was that people may have a valid use for this, and as long as they know what they're getting into the library doesn't have to remind them on every call.

Another practical concern would be what happens when someone tries to call `cu(large state tree)`. Not setting `maxlog` would mean other logging is drowned out because this warning would trigger for every `Leaf`.

Reply:
I agree that printing 100 times on a big model is too much. Ideal would be once, on every invocation, IMO... but that's hard to make happen.

It's not the world's biggest correctness bug to ignore shared parameters, so maybe we should live with it. Maybe the message should say that's what the problem is?